### PR TITLE
Add market service

### PR DESCRIPTION
**Pull request by chilianyi**

### CODE REVIEW

File: api/market.proto [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    google.protobuf.StringValue status = 3;
+    google.protobuf.StringValue description = 4;
+}
+
+message CreateMarketResponse {
+    google.protobuf.StringValue market_id = 1;
+}
+
+message DescribeMarketsRequest {
+    repeated string market_id = 1;
```
Comment (zheng1) [inline api/market.proto]:
```
1-10 reversed params
11+ custom params
```

File: api/market.proto [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+message DescribeMarketUsersRequest {
+    repeated string market_id = 1;
+    repeated string user_id = 2;
+    uint32 limit = 3;
+    uint32 offset = 4;
+    google.protobuf.StringValue search_word = 5;
+    google.protobuf.StringValue sort_key = 6;
+    google.protobuf.BoolValue reverse = 7;
```
Comment (zheng1) [inline api/market.proto]: ditto

File: api/market.proto [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    	option (google.api.http) = {
+    		post: "/v1/markets:leave"
+    	};
+    }
+    rpc DescribeMarketUsers (DescribeMarketUsersRequest) returns (DescribeMarketUsersResponse) {
+    	option (grpc.gateway.protoc_gen_swagger.options.openapiv2_operation) = {
+    		summary: "describe users with filter"
+    	};
+    	option (google.api.http) = {
+    		get: "/v1/market_user"
```
Comment (zheng1) [inline api/market.proto]: `market_users`

File: pkg/db/schema/market/V0_1__init.sql [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+	ON market (create_time);
+CREATE INDEX market_status_time_idx
+	ON market (status_time);
+
+CREATE TABLE IF NOT EXISTS market_user (
+	market_id   VARCHAR(50) NOT NULL,
+	user_id     VARCHAR(50) NOT NULL,
+	owner       VARCHAR(50) NOT NULL,
+	create_time TIMESTAMP   NOT NULL DEFAULT CURRENT_TIMESTAMP,
+	PRIMARY KEY (market_id)
```
Comment (zheng1) [inline pkg/db/schema/market/V0_1__init.sql]: `PRIMARY KEY (market_id, user_id)`

File: pkg/service/market/checker.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	"context"
+
+	"openpitrix.io/openpitrix/pkg/manager"
+	"openpitrix.io/openpitrix/pkg/pb"
+)
+
+func (s *Server) Checker(ctx context.Context, req interface{}) error {
+	switch r := req.(type) {
+	case *pb.CreateMarketRequest:
+		return manager.NewChecker(ctx, r).Required("name", "visibility", "status").Exec()
```
Comment (zheng1) [inline pkg/service/market/checker.go]: remove `status`

File: pkg/service/market/checker.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+		return manager.NewChecker(ctx, r).Required("market_id").Exec()
+
+	case *pb.UserJoinMarketRequest:
+		return manager.NewChecker(ctx, r).Required("market_id", "user_id").Exec()
+
+	case *pb.UserLeaveMarketRequest:
+		return manager.NewChecker(ctx, r).Required("market_id", "user_id").Exec()
+
+	case *pb.DescribeMarketUsersRequest:
+		return manager.NewChecker(ctx, r).Required("market_id").Exec()
```
Comment (zheng1) [inline pkg/service/market/checker.go]: remove `.Required("market_id")`

File: pkg/service/market/handler.go [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+	"openpitrix.io/openpitrix/pkg/util/senderutil"
+)
+
+func (p *Server) CreateMarket(ctx context.Context, req *pb.CreateMarketRequest) (*pb.CreateMarketResponse, error) {
+	s := senderutil.GetSenderFromContext(ctx)
+
+	newMarket := models.NewMarket(
+		req.GetName().GetValue(),
+		req.GetVisibility().GetValue(),
+		req.GetStatus().GetValue(),
```
Comment (zheng1) [inline pkg/service/market/handler.go]: remove `req.GetStatus().GetValue(),`

File: pkg/service/market/handler.go [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+	offset := pbutil.GetOffsetFromRequest(req)
+	limit := pbutil.GetLimitFromRequest(req)
+
+	query := pi.Global().DB(ctx).
+		Select(models.MarketColumns...).
+		From(constants.TableMarket).
+		Offset(offset).
+		Limit(limit).
+		Where(manager.BuildFilterConditions(req, constants.TableMarket))
+
```
Comment (zheng1) [inline pkg/service/market/handler.go]: SortKey

File: pkg/service/market/checker.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+	}
+	return nil
+}
```
Comment (zheng1) [inline pkg/service/market/checker.go]: Add `Builder`

File: pkg/service/market/handler.go [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+		TotalCount: uint32(count),
+	}
+
+	return res, nil
+}
+
+func (p *Server) ModifyMarket(ctx context.Context, req *pb.ModifyMarketRequest) (*pb.ModifyMarketResponse, error) {
+	marketId := req.GetMarketId().GetValue()
+
+	_, err := p.getMarket(ctx, marketId)
```
Comment (zheng1) [inline pkg/service/market/handler.go]: CheckPermission

File: pkg/service/market/checker.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	"openpitrix.io/openpitrix/pkg/pb"
+)
+
+func (s *Server) Checker(ctx context.Context, req interface{}) error {
+	switch r := req.(type) {
+	case *pb.CreateMarketRequest:
+		return manager.NewChecker(ctx, r).Required("name", "visibility", "status").Exec()
+
+	case *pb.ModifyMarketRequest:
+		return manager.NewChecker(ctx, r).Required("market_id").Exec()
```
Comment (zheng1) [inline pkg/service/market/checker.go]: check status must be in [enabled, disabled]

File: pkg/service/market/handler.go (hunk 10 above)
Comment (zheng1) [reply pkg/service/market/handler.go]: Check status cannot be `deleted`

File: pkg/service/market/handler.go [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorModifyResourceFailed)
+	}
+
+	attributes := manager.BuildUpdateAttributes(req,
+		constants.ColumnName, constants.ColumnVisibility, constants.ColumnStatus, constants.ColumnDescription)
+	attributes[constants.ColumnStatusTime] = time.Now()
```
Comment (zheng1) [inline pkg/service/market/handler.go]: if status is changed, update status_time

File: pkg/service/market/handler.go [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+		MarketId: req.GetMarketId(),
+	}
+
+	return res, nil
+}
+
+func (p *Server) DeleteMarkets(ctx context.Context, req *pb.DeleteMarketsRequest) (*pb.DeleteMarketsResponse, error) {
+	marketIds := req.GetMarketId()
+
+	_, err := CheckMarketsPermission(ctx, marketIds)
```
Comment (zheng1) [inline pkg/service/market/handler.go]: check status cannot be deleted

File: pkg/service/market/handler.go [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+	userIds := req.GetUserId()
+	marketIds := req.GetMarketId()
+
+	_, err := CheckMarketsPermission(ctx, marketIds)
+	if err != nil {
+		return nil, err
+	}
+
+	var marketUsers []*models.MarketUser
+	markets, err := p.getMarkets(ctx, marketIds)
```
Comment (zheng1) [inline pkg/service/market/handler.go]: delete `p.getMarkets`

File: pkg/service/market/handler.go [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorInternalError)
+	}
+
+	for _, market := range markets {
+		for _, userid := range userIds {
```
Comment (zheng1) [inline pkg/service/market/handler.go]: `userid` => `userId`

File: pkg/service/market/handler.go [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+			marketUsers = append(marketUsers, models.NewMarketUser(market.MarketId,
+				userid, market.Owner))
+		}
+	}
+
+	_, err = pi.Global().DB(ctx).
```
Comment (zheng1) [inline pkg/service/market/handler.go]: for loop with `marketUsers`

Comment (zheng1) [reply pkg/service/market/handler.go]:
```go
count, err := DB(ctx).Select().From(market_user).Where(db.Eq(user_id, uid)).Where(db.Eq(market_id, mid)).Count()
```

File: pkg/service/market/handler.go [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+	_, err := pi.Global().DB(ctx).
+		InsertInto(constants.TableMarket).
+		Record(newMarket).
+		Exec()
+
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorCreateResourceFailed)
+	}
+
+	res := &pb.CreateMarketResponse{
```
Comment (zheng1) [inline pkg/service/market/handler.go]: owner join market_user

File: pkg/service/market/handler.go [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+		MarketId: req.GetMarketId(),
+		UserId:   req.GetUserId(),
+	}
+
+	return res, nil
+}
+
+func (p *Server) DescribeMarketUsers(ctx context.Context, req *pb.DescribeMarketUsersRequest) (*pb.DescribeMarketUsersResponse, error) {
+	var marketUsers []*models.MarketUser
+
```
Comment (zheng1) [inline pkg/service/market/handler.go]: if developer:
```go
if len(req.GetOwner()) > 0 {
  req.Owner = []string{s.UserId}
} else {
  req.UserId = []string{s.UserId}
}
```
if user:
```
req.UserId = []string{s.UserId}
```

File: pkg/service/market/server.go [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+type Server struct {
+}
+
+func Serve(cfg *config.Config) {
+	pi.SetGlobal(cfg)
+	s := Server{}
+	manager.NewGrpcServer("market-manager", constants.MarketManagerPort).
+		ShowErrorCause(cfg.Grpc.ShowErrorCause).
+		WithChecker(s.Checker).
```
Comment (zheng1) [inline pkg/service/market/server.go]:
```
WithBuilder(s.Builder).
```

File: Makefile [hunk 20]
```diff
@@ -28,8 +28,8 @@ define get_build_flags
 		-X $(TRAG.Version).BuildDate="$(DATE)")
 endef
 
-COMPOSE_APP_SERVICES=openpitrix-runtime-manager openpitrix-app-manager openpitrix-category-manager openpitrix-repo-indexer openpitrix-api-gateway openpitrix-repo-manager openpitrix-job-manager openpitrix-task-manager openpitrix-cluster-manager openpitrix-pilot-service openpitrix-iam-service
-COMPOSE_DB_CTRL=openpitrix-app-db-ctrl openpitrix-repo-db-ctrl openpitrix-runtime-db-ctrl openpitrix-job-db-ctrl openpitrix-task-db-ctrl openpitrix-cluster-db-ctrl openpitrix-iam-db-ctrl
+COMPOSE_APP_SERVICES=openpitrix-runtime-manager openpitrix-app-manager openpitrix-category-manager openpitrix-repo-indexer openpitrix-api-gateway openpitrix-repo-manager openpitrix-job-manager openpitrix-task-manager openpitrix-cluster-manager openpitrix-pilot-service openpitrix-iam-service openpitrix-market-service
```
Comment (zheng1) [inline Makefile]: `openpitrix-market-manager`

File: pkg/config/init_global_config.go [hunk 21]
```diff
@@ -27,6 +27,7 @@ repo:
 cluster:
   frontgate_conf: '{"app_id":"app-ABCDEFGHIJKLMNOPQRST","version_id":"appv-ABCDEFGHIJKLMNOPQRST","name":"frontgate","description":"OpenPitrixbuilt-infrontgateservice","subnet":"","nodes":[{"container":{"type":"docker","image":"openpitrix/openpitrix:metadata"},"count":1,"cpu":1,"memory":1024,"volume":{"size":10,"mount_point":"/data","filesystem":"ext4"}}]}'
   frontgate_auto_delete: true
+  frontgate_auto_update: false
```
Comment (zheng1) [inline pkg/config/init_global_config.go]: It might need to be deleted, not related to this PR

File: pkg/service/market/handler.go (hunk 12 above)
Comment (zheng1) [reply pkg/service/market/handler.go]: should be fixed

File: pkg/service/market/handler.go [hunk 22]
```diff
@@ -0,0 +1,255 @@
...
+			}
+
+			if count == 0 {
+				marketUsers = append(marketUsers, models.NewMarketUser(market.MarketId,
+					userId, market.Owner))
+			}
+		}
+	}
+
+	_, err = pi.Global().DB(ctx).
```
Comment (zheng1) [inline pkg/service/market/handler.go]: should check `len(marketUsers) == 0`

File: pkg/service/market/permission.go [hunk 23]
```diff
@@ -0,0 +1,74 @@
+package market
```
Comment (zheng1) [inline pkg/service/market/permission.go]: This file should be generated

File: test/market/market_test.go [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+	describeMarketsParams.SetMarketID([]string{marketId1})
+	describeMarketsResp, err := client.MarketManager.DescribeMarkets(describeMarketsParams, nil)
+	require.NoError(t, err)
+
+	markets := describeMarketsResp.Payload.MarketSet
+	require.Equal(t, 1, len(markets))
+
+	market := markets[0]
+
+	require.Equal(t, testMarketName2, market.Name)
```
Comment (zheng1) [inline test/market/market_test.go]: should test `Status | Visibility  | Description` of  the market record

File: pkg/config/init_global_config.go (hunk 21 above)
Comment (zhou1203) [reply pkg/config/init_global_config.go]: It have come to exist in openpitrix, i didn't add it

File: Makefile [hunk 25]
```diff
@@ -28,8 +28,8 @@ define get_build_flags
 		-X $(TRAG.Version).BuildDate="$(DATE)")
 endef
 
-COMPOSE_APP_SERVICES=openpitrix-runtime-manager openpitrix-app-manager openpitrix-category-manager openpitrix-repo-indexer openpitrix-api-gateway openpitrix-repo-manager openpitrix-job-manager openpitrix-task-manager openpitrix-cluster-manager openpitrix-pilot-service openpitrix-iam-service
-COMPOSE_DB_CTRL=openpitrix-app-db-ctrl openpitrix-repo-db-ctrl openpitrix-runtime-db-ctrl openpitrix-job-db-ctrl openpitrix-task-db-ctrl openpitrix-cluster-db-ctrl openpitrix-iam-db-ctrl
+COMPOSE_APP_SERVICES=openpitrix-runtime-manager openpitrix-app-manager openpitrix-category-manager openpitrix-repo-indexer openpitrix-api-gateway openpitrix-repo-manager openpitrix-job-manager openpitrix-task-manager openpitrix-cluster-manager openpitrix-market-manager openpitrix-pilot-service openpitrix-iam-service openpitrix-market-service
```
Comment (zheng1) [inline Makefile]: `openpitrix-market-service` should delete

**Comment by zheng1** [comment]
Great work. Thanks for your contribution.

**Pull request merged**